### ISSUE TITLE
PLEASE HELP WD MyBook Essentials 2TB stopped mounting

### ISSUE THREAD

**Issue by mrthielo** [andlabs/reallymine#63]
Hi there,

 I would really like to beg for some help here... My external WD MyBook Essentials 2TB stopped working  with all my work and personal data files. I tried to read it in another SATA enclosure both in Linux and Windows before knowing it is encrypted, of course without any success.
 I did not format or messed with anything on it, and followed every step on the tutorial "Mounting encrypted WD drives in Linux" and everything was ok until the line (please noticed that I changed sdc to sdb)

```
cat dek.hex | xxd -p -r | sudo cryptsetup -d - --hash=plain \
--key-size=256 -c aes-ecb create wd /dev/sdb
```

where it seems to stay forever waiting for some passphrase (even if I type anything). I have set no password to my disk. My chip is the Symwave 6316-3VB14.

My kb.bin dump is

```
00000000  53 59 4d 57 f8 01 9a b7  00 00 00 00 00 00 00 02  |SYMW............|
00000010  e8 8b a6 29 c9 be de 3c  a4 9c cf 69 8a 16 d8 ad  |...)...<...i....|
00000020  e4 27 22 b2 e5 ea 4a e8  cc 54 b1 da ee 95 7d 1a  |.'"...J..T....}.|
00000030  42 ea 58 26 76 70 69 34  b5 b9 88 15 20 03 b1 72  |B.X&vpi4.... ..r|
00000040  88 eb 65 f3 e7 f9 70 91  cb ee 9a 09 97 ad 05 58  |..e...p........X|
00000050  6d b3 6e e3 cd c9 78 5f  c0 85 cb fe 8d 06 50 43  |m.n...x_......PC|
00000060  6e 50 b6 0f 8c 30 36 1a  fa 9b 25 8e 6a 6b 26 32  |nP...06...%.jk&2|
00000070  61 72 02 04 65 f3 a9 c0  55 b5 b4 a1 c7 e7 d4 0c  |ar..e...U.......|
00000080  f2 3b 52 f1 69 e8 b3 46  00 00 00 00 00 00 00 00  |.;R.i..F........|
00000090  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000001c0  e8 df 88 00 00 00 00 00  00 00 02 00 00 00 00 00  |................|
000001d0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000200

```
(after this I performed the correction for the little-endianess as well)

and my kek.hex contents is 

`03141592653589793238462643383279fcebea6d9aca7686cdc7b9d9bcc7cd86`

My disk model is WD20EARS. 

What am I doing wrong? 

Thank you a lot guys. You are my last hope, I am really desperate!

Marcelo

**Comment by themaddoctor** [comment]
That happens when xxd (part of vim) is not installed. You need to install it.

BTW, you posted in the wrong project.

**Comment by athomic1** [comment]
I see your kb.bin dump. Question: Is this what it looked like BEFORE you applied the endianness correction? If so, I DON'T think you need to correct your endianness, here. If it were backwards, the first four characters should have been dumped as WMYS, not SYMW. I believe it was the scrambling that tipped me off to the situation when I was still working to recover my own drive.

Again, IF your sector is coming out exactly as above BEFORE you correct endianness, you probably do NOT need that correction. If you haven't already, I would suggest you give it a try WITHOUT the correction, and see where that gets you.

One other note: I don't really know how the program interacts with the user (I was able to perform my own extraction through other means, just using information from this project), but keep in mind the extraction of 2TB of data takes a LONG time, from a dozen hours, to a day or longer. Unless the program normally displays some sort of progress bar or something, it might just sit there until the task is completed. One way to try and keep an eye on its activity is to start another terminal window, and run 'top' in that. If the process IS running, you should see the command somewhere in the display, which will probably bounce around a bit, depending on what else is going on. Other than that, I'm not sure what else to suggest, but it sounds, from what you've said, you should still have a pretty good chance of recovering your data. Hopefully, it's just the endianness correction, like I said above.

Keep us posted, and best of luck!

**Comment by athomic1** [comment]
Wait, wrong project? Did he post this somewhere else initially, or am I responding in the wrong place?

**Comment by themaddoctor** [comment]
"Mounting ......pdf" is from the linux-mybook-tools project. This is the reallymine project.

It doesn't really matter. If you install vim (with xxd), you are probably set.
sudo apt-get install vim

**Comment by themaddoctor** [comment]
Is this your DEK?
bdc506c40ddd8ac723f19e505cafa49a0ccce274b8e723b5179dbf2f897f4b96

**Comment by mrthielo** [comment]
@themaddoctor: Sorry for having posted in the wrong thread, I didnt' realize it. Maybe due to the panic state I am currently in.

I typed xxd --help and it is actually installed. Does it need to be some specific version? vim is also there. What could be wrong?

```
cat dek.hex 
bdc506c40ddd8ac723f19e505cafa49a0ccce274b8e723b5179dbf2f897f4b96

```
Yes, it seems to be my DEK.

What does that mean? 

@athomic1:

I am going to give a try to the kb.bin without reversing it and get back to you on this.

BTW, is this normal for an encrypted disk or...?

```
$ sudo fdisk -l /dev/sdb

Disk /dev/sdb: 2000.4 GB, 2000398934016 bytes
255 heads, 63 sectors/track, 243201 cylinders, total 3907029168 sectors
Units = sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disk identifier: 0x221b4317

Disk /dev/sdb doesn't contain a valid partition table

```

Big thanks for the help,

Marcelo

**Comment by mrthielo** [comment]
I almost forgot,

The prompt got stuck as I mentioned, and I took a look at _top_ already. Nothing related appears there, as disk activity also is not perceivable in /dev/sdb. 
It looks bizarre that even ctrl+c can't stop it the cat line. If I press ctrl+z the process stops. When I close the terminal, is seems to go away, but -
After that, I looked at /proc/partitions and there is a new thing there:
 252        0 1953514584 dm-0
But any reference to /dev/mapper/wd I try returns 
 ERROR: cannot open `/dev/mapper/wd' (No such file or directory)

**Comment by themaddoctor** [comment]
Maybe /dev/dm-0 is your decrypted device.

If you copied and pasted from the PDF, then you would have included a tab character that you should not have. Try again by typing the command, all on one line, without the \

**Comment by themaddoctor** [comment]
Another reason that command can hang is if cryptsetup is not installed.

**Comment by mrthielo** [comment]
Hello,

I have checked and cryptsetup is already there. On the other hand, I already copied and pasted from the pdf in gedit, turned it into one single line, removed the \ and still the same behavior. Maybe it was the tab character there, going to type it as soon as I can reboot (I am working now). When it works, should it do the actual decryption that will take so much time for a great volume or should it just stand there waiting for accesses in the volume to do the decryption?
I will repeat the process without the reversing. I am just afraid of waiting because I haven't seen any hard drive activity if it should happen. Hope these new hints with change it.

Thank you,

Marcelo

**Comment by themaddoctor** [comment]
The cryptsetup command should take less than a second. If it succeeds, then you mount. Then how long it takes depends on how much you want to copy.

Another way to go is to use the DEK with reallymine and make a decrypted copy of the disk.

**Comment by mrthielo** [comment]
I think that I will keep trying to mount the disk and, when I get success, I will try reallymine to make a full copy of it to another media. I am still trying, hope to return and report results soon.

Regards,
Marcelo

**Comment by mrthielo** [comment]
Hi, I'm back. Typed this as a single line, no tab characters

`$ cat dek.hex | xxd -p -r | sudo cryptsetup -d - --hash=plain --key-size=256 -c aes-ecb create wd /dev/sdb`

and the result is the same (still stalled in shell).
peeking at /proc/partitions I have noticed that dm-0 appeared again:

  ```
 8        0  976762584 sda
   8        1   19998720 sda1
   8        2          1 sda2
   8        3  819105792 sda3
   8        4   97655808 sda4
   8        5   39999488 sda5
  11        0    1048575 sr0
   8       16 1953514584 sdb
 252        0 1953514584 dm-0
```
but nothing but `control` is found under `/dev/mapper`. 

I can also see

```
$ lsblk
NAME        MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINT
sda           8:0    0 931.5G  0 disk  
├─sda1        8:1    0  19.1G  0 part  [SWAP]
├─sda2        8:2    0     1K  0 part  
├─sda3        8:3    0 781.2G  0 part  /home
├─sda4        8:4    0  93.1G  0 part  /windows
└─sda5        8:5    0  38.2G  0 part  /
sdb           8:16   0   1.8T  0 disk  
└─wd (dm-0) 252:0    0   1.8T  0 crypt 
sr0          11:0    1  1024M  0 rom   
```
But I don't seem to be able to access anything, probably because the command could not complete.
Maybe it is a hardware failure? Shall I send my disk to a data recovery specialist?

Regards,

Marcelo

**Comment by themaddoctor** [comment]
Yeah, so isn't /dev/dm-0 just a link to /dev/mapper/wd?

**Comment by themaddoctor** [comment]
I don't think it's hardware, because the cryptsetup command doesn't do any probing or anything. It just sets up a filter.

It will hang until it receives enough bits for the key.

Anyway, I did this
```
echo  bdc506c40ddd8ac723f19e505cafa49a0ccce274b8e723b5179dbf2f897f4b96|xxd -p -r | sudo cryptsetup -d - --hash=plain --key-size=256 -c aes-ecb create wd somefile.bin
```
and it only took one second.

**Comment by themaddoctor** [comment]
I don't think it's hardware, because the cryptsetup command doesn't do any probing or anything. It just sets up a filter.

It will hang until it receives enough bits for the key.

Anyway, I did this
```
echo  bdc506c40ddd8ac723f19e505cafa49a0ccce274b8e723b5179dbf2f897f4b96|xxd -p -r | sudo cryptsetup -d - --hash=plain --key-size=256 -c aes-ecb create wd somefile.bin
```
and it only took one second.

**Comment by mrthielo** [comment]
I just repeated the line in the tutorial using /dev/sdg (random usb stick) instead of /dev/sdb and it worked pretty fast. Will keep trying, hope I can stop bothering you soon.

Cheers

**Comment by mrthielo** [comment]
Hello again,

Just tried again, still no success. But this time, I passed the key --debug to cryptsetup to see what is going on. It seems like it stops "waiting for a zero that never comes".
How severe this seems to be?

```
 cat dek.hex | xxd -p -r | sudo cryptsetup -d - --hash=plain --key-size=256 --debug -c aes-ecb create wd /dev/sdb
[sudo] password for thielo: 
# cryptsetup 1.6.1 processing "cryptsetup -d - --hash=plain --key-size=256 --debug -c aes-ecb create wd /dev/sdb"
# Running command open.
# Locking memory.
# Installing SIGINT/SIGTERM handler.
# Unblocking interruption on signal.
# Allocating crypt device /dev/sdb context.
# Trying to open and read device /dev/sdb.
# Initialising device-mapper backend library.
# Timeout set to 0 miliseconds.
# Password retry count set to 3.
# Formatting device /dev/sdb as type PLAIN.
# Crypto backend (gcrypt 1.5.3) initialized.
# Activating volume wd [keyslot -1] using keyfile -.
# dm version   OF   [16384] (*1)
# dm versions   OF   [16384] (*1)
# Detected dm-crypt version 1.13.0, dm-ioctl version 4.29.0.
# Device-mapper backend running with UDEV support enabled.
# dm status wd  OF   [16384] (*1)
# STDIN descriptor passphrase entry requested.
# Calculated device size is 3907029168 sectors (RW), offset 0.
# Trying to activate PLAIN device wd using cipher aes-ecb.
# DM-UUID is CRYPT-PLAIN-wd
# Udev cookie 0xd4deabb (semid 32768) created
# Udev cookie 0xd4deabb (semid 32768) incremented to 1
# Udev cookie 0xd4deabb (semid 32768) incremented to 2
# Udev cookie 0xd4deabb (semid 32768) assigned to CREATE task(0) with flags (0x0)
# dm create wd CRYPT-PLAIN-wd OF   [16384] (*1)
# dm reload wd  OFW    [16384] (*1)
# dm resume wd  OFW    [16384] (*1)
# wd: Stacking NODE_ADD (252,0) 0:6 0660 [verify_udev]
# wd: Stacking NODE_READ_AHEAD 256 (flags=1)
# Udev cookie 0xd4deabb (semid 32768) decremented to 1
# Udev cookie 0xd4deabb (semid 32768) waiting for zero
```

**Comment by themaddoctor** [comment]
Mine says
# Device-mapper backend running with UDEV support disabled.

**Comment by themaddoctor** [comment]
Someone else had a similar problem:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=781468

**Comment by mrthielo** [comment]
Hi, I'm back! I came to report some success and ask for more help...

After disabling UDEV using the line

`sudo /etc/init.d/udev stop`

I managed to go through the tutorial and mount the virtual device. It had the NTFS corrupted and when mounting the loopback device he told me he was fixing that. That took a bit, since the disk is 2TB and has no partitions. Now I am able to see all the names of the files in the root directory. I can also enter any of the directories there but when I try to access any file in the root folder or any subdirectory I get the error message:

`ls: reading directory Work: Input/output error`

What shall I do? Maybe run a fsck there?

Thank you! Now I have some hope to get my files back...
